vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,743 @@
+      index_queue.push(std::make_pair(priority, index));
+    }
+
+    for (auto& score : scores)
+    {
+      score.live_interactions.clear();
+      interaction_vec v = oc.gen_interactions(ns_counter, configs[score.config_index].exclusions);
+      score.live_interactions = v;
+    }
+  }
+  else
+  {
+    // Save scores
+    score_size = scores.size();
+    msg << "score_size " << score_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&score_size), sizeof(score_size), "", read, msg, text);
+    for (auto& score : scores) { score.save_load_scored_config(model_file, read, text); }
+
+    // Save configs
+    config_size = configs.size();
+    msg << "config_size " << config_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&config_size), sizeof(config_size), "", read, msg, text);
+    for (auto& ind_config : configs)
+    {
+      bin_text_read_write_fixed(model_file, (char*)&ind_config.first, sizeof(ind_config.first), "", read, msg, text);
+      ind_config.second.save_load_exclusion_config(model_file, read, text);
+    }
+
+    // Save ns_counter
+    ns_counter_size = ns_counter.size();
+    msg << "ns_counter_size " << ns_counter_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&ns_counter_size), sizeof(ns_counter_size), "", read, msg, text);
+    for (auto& ns_count : ns_counter)
+    {
+      bin_text_read_write_fixed(model_file, (char*)&ns_count.first, sizeof(ns_count.first), "", read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&ns_count.second), sizeof(ns_count.second), "", read, msg, text);
+    }
+
+    // Save index_queue
+    index_queue_size = index_queue.size();
+    msg << "index_queue_size " << index_queue_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&index_queue_size), sizeof(index_queue_size), "", read, msg, text);
+    while (!index_queue.empty())
+    {
+      auto& pri_ind = index_queue.top();
+      bin_text_read_write_fixed(model_file, (char*)&pri_ind.first, sizeof(pri_ind.first), "", read, msg, text);
+      bin_text_read_write_fixed(model_file, (char*)&pri_ind.second, sizeof(pri_ind.second), "", read, msg, text);
+      index_queue.pop();
+    }
+  }
+}
+
+void config_manager::persist(metric_sink& metrics)
+{
+  metrics.int_metrics_list.emplace_back("test_county", county);
+  metrics.int_metrics_list.emplace_back("current_champ", current_champ);
+  for (size_t stride = 0; stride < scores.size(); ++stride)
+  { scores[stride].persist(metrics, "_" + std::to_string(stride)); }
+}
+
+// This sets up example with correct ineractions vector
+void config_manager::apply_config(example* ec, size_t stride)
+{
+  if (ec == nullptr) { return; }
+  if (stride < max_live_configs) { ec->interactions = &(scores[stride].live_interactions); }
```

Comment:
Are wildcard interactions ever used? I imagine not?

How does this reduction interact with CCB? Since it modifies the interactions state?

Reply:
I was not considering CCB compat for MVP. Right now I'm failing on setup if ccb is enabled.

Reply:
We do have a pending task to audit all those reductions flagged as possibly modifying interaction state.